openeo/internal/jupyter.py

```diff
@@ -0,0 +1,70 @@
+from string import Template
+from collections import Mapping
+import json
+
+# These classes are proxies to visualize openEO responses nicely in Jupyter
+# To show the actual list or dict in Jupyter, use repr() or print()
+
+SCRIPT_URL = 'https://cdn.jsdelivr.net/npm/@openeo/vue-components@2.0.0-beta.1/assets/openeo.js'
+
+class JupyterIntegration():
+
+	# Class instance to store whether the script has been inserted into the DOM
+	# TODO: This fails if a page is reloaded in Jupyter. The variable is still True, but the script doesn't exist in the page any longer.
+	script_sent = False
+
+	def __init__(self, component: str, parameters: dict = {}):
+		self.component = component
+		self.parameters = parameters
+
+	def _repr_html_(self):
+		# Output script tag only once, otherwise JS throws errors
+		script = ''
+		if JupyterIntegration.script_sent == False:
```

Comment:
Can't this check whether the script has already been included already be moved to JS space?
The JS frontend and Python backend work separately and I think you can get into situation where Python space `script_sent`  is out of sync with the reality of the JS-app side. For example: if you already included javascript in the JS app this way, and then restart your python kernel, the python space `script_sent`  would be reset to False, which is wrong.

Reply:
I'd love to, but I'm not sure how I can access the Jupyter DOM from the Python client. There are definitely cases where this could fail as documented in the code, especially a page reload. Kernel restart is less of a problem as it still works although some errors are logged to the JS console. With page reload it doesn't display anything if the JS is missing, although I could try to at least print a message that says what to do.

Reply:
Maybe I can somehow build a JS wrapper script that is executed each time...

Reply:
@soxofaan 


Do you have a clue why this 
```py

        return """
        <script type="text/javascript">
        if (!window.customElements.get('openeo-{component}')) {
            var el = document.createElement('script');
            el.src = "{script}";
            document.appendChild(el);
        }
        </script>
        <openeo-{component}>
            <script type="application/json">{props}</script>
        </openeo-{component}>
        """.format(
            script = SCRIPT_URL,
            component = self.component,
            props = json.dumps(self.parameters)
        )
```
gives a "Invalid format string" error? If I remove the "{script}" it works though, but how to fill it then?

Reply:
because it trips over the curly braces of the `if () {... } `  

you can escape these with double braces: 
    
    >>> "{x} {{y}}".format(x="foo", y="bar")
    'foo {y}'


But if you indeed want to use more advanced javascript in that string, it might be easier to return to that `string.Template` :rofl:

Reply:
> how I can access the Jupyter DOM from the Python client. 

> a JS wrapper script that is executed each time..

yes, I meant something like a quick JS-side checking wrapper

Reply:
Wrote a small JS script that checks the existence of the required HTML tags and loads the script only if it's missing. It seems to work in all cases I tested.